faction.cpp

```diff
@@ -301,144 +310,366 @@ AString Faction::FactionTypeStr()
+	GmData data = collect_gm_data();
 
-	if (IsNPC() && num == 1) {
-		if (Globals->GM_REPORT || (pGame->month == 0 && pGame->year == 1)) {
-			int i, j;
-			// Put all skills, items and objects in the GM report
-			shows.DeleteAll();
-			for (i = 0; i < NSKILLS; i++) {
-				for (j = 1; j < 6; j++) {
-					shows.Add(new ShowSkill(i, j));
-				}
-			}
-			if (shows.Num()) {
-				f << "Skill reports:\n";
-				forlist(&shows) {
-					AString *string = ((ShowSkill *)elem)->Report(this);
-					if (string) {
-						f << '\n' << string->const_str() << '\n';
-						delete string;
-					}
-				}
-				shows.DeleteAll();
-				f << '\n';
-			}
+	bool need_header = true;
+	for (auto &skillshow : data.skills) {
+		if(need_header) { f << "Skill reports:\n"; need_header = false; }
+		AString *string = skillshow.Report(this);
+		if (string) {
+			f << '\n' << string->const_str() << '\n';
+			delete string;
+		}
+	}
+	if(!need_header) f << '\n';
 
-			itemshows.DeleteAll();
-			for (i = 0; i < NITEMS; i++) {
-				AString *show = ItemDescription(i, 1);
-				if (show) {
-					itemshows.Add(show);
-				}
-			}
-			if (itemshows.Num()) {
-				f << "Item reports:\n";
-				forlist(&itemshows) {
-					f << '\n' << ((AString *)elem)->const_str() << '\n';
-				}
-				itemshows.DeleteAll();
-				f << '\n';
-			}
+	need_header = true;
+	for(auto &itemshow : data.items) {
+		if(need_header) { f << "Item reports:\n"; need_header = false; }
+			f << '\n' << itemshow << '\n';
+	}
+	if (!need_header) f << '\n';
 
-			objectshows.DeleteAll();
-			for (i = 0; i < NOBJECTS; i++) {
-				AString *show = ObjectDescription(i);
-				if (show) {
-					objectshows.Add(show);
-				}
-			}
-			if (objectshows.Num()) {
-				f << "Object reports:\n";
-				forlist(&objectshows) {
-					f << '\n' << ((AString *)elem)->const_str() << '\n';
-				}
-				objectshows.DeleteAll();
-				f << '\n';
-			}
+	need_header = true;
+	for(auto &objectshow : data.objects) {
+		if(need_header) { f << "Object reports:\n"; need_header = false; }
+		f << '\n' << objectshow << '\n';
+	}
+	if (!need_header) f << '\n';
+
+	present_regions.DeleteAll();
+	forlist(&(game->regions)) {
+		ARegion *reg = (ARegion *)elem;
+		ARegionPtr *ptr = new ARegionPtr;
+		ptr->ptr = reg;
+		present_regions.Add(ptr);
+	}
+	{
+		// extra block because of foreach macro.
+		forlist(&present_regions) {
+			((ARegionPtr*)elem)->ptr->WriteReport(f, this, game->month, &(game->regions));
+		}
+	}
+	present_regions.DeleteAll();
 
-			present_regions.DeleteAll();
-			forlist(&(pGame->regions)) {
-				ARegion *reg = (ARegion *)elem;
-				ARegionPtr *ptr = new ARegionPtr;
-				ptr->ptr = reg;
-				present_regions.Add(ptr);
-			}
-			{
-				forlist(&present_regions) {
-					((ARegionPtr*)elem)->ptr->WriteReport(f, this, pGame->month, &(pGame->regions));
-				}
-			}
-			present_regions.DeleteAll();
+	errors.DeleteAll();
+	events.DeleteAll();
+	battles.DeleteAll();
+}
+
+void Faction::write_json_gm_report(json& j, Game *game) {
```

Comment:
the gm reports were taking up a lot of space and making the report function hard to read, so I pulled all the gm stuff out into it's own pair of functions.   I also just collect all the data up front and output it.  This made the code cleaner imho.